From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wechandoit <ethanchan7981@gmail.com>
Date: Wed, 3 Dec 2025 21:03:11 -0500
Subject: [PATCH] Saturn: Avoid unnecessary object creation


diff --git a/com/mojang/datafixers/functions/PointFree.java b/com/mojang/datafixers/functions/PointFree.java
index 81b0dff25809f0c8f2c7a6ed8e6d19fe4b0d6f05..3f6a05c177d1bb5567aff198e55895ad1a292bec 100644
--- a/com/mojang/datafixers/functions/PointFree.java
+++ b/com/mojang/datafixers/functions/PointFree.java
@@ -19,12 +19,11 @@ public abstract class PointFree<T> {
         if (!initialized) {
             synchronized (this) {
                 if (!initialized) {
-                    value = eval();
                     initialized = true;
                 }
             }
         }
-        return value;
+        return eval(); // Paper - Saturn: Avoid unnecessary object creation
     }
 
     public abstract Type<T> type();
diff --git a/com/mojang/datafixers/types/Type.java b/com/mojang/datafixers/types/Type.java
index b05a8f5f1da64d8b821a1423e865bdb4bc764694..4ef6f481d1a418c38e0afb50365950938d0e3e8d 100644
--- a/com/mojang/datafixers/types/Type.java
+++ b/com/mojang/datafixers/types/Type.java
@@ -93,10 +93,7 @@ public abstract class Type<A> implements App<Type.Mu, A> {
     }
 
     public TypeTemplate template() {
-        if (template == null) {
-            template = buildTemplate();
-        }
-        return template;
+        return buildTemplate(); // Paper - Saturn: Avoid unnecessary object creation
     }
 
     public abstract TypeTemplate buildTemplate();
@@ -114,10 +111,7 @@ public abstract class Type<A> implements App<Type.Mu, A> {
     }
 
     public final Codec<A> codec() {
-        if (codec == null) {
-            codec = buildCodec();
-        }
-        return codec;
+        return buildCodec(); // Paper - Saturn: Avoid unnecessary object creation
     }
 
     protected abstract Codec<A> buildCodec();
