From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wechandoit <ethanchan7981@gmail.com>
Date: Thu, 7 May 2026 00:11:13 -0400
Subject: [PATCH] Jettpack: Reduce array allocations


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
index 3f5bd8c914df5a1fda812c9286647839c69c1cd8..c8864ad1322d2dea9e8b0398195cada6949aefd8 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
@@ -30,6 +30,7 @@ import net.minecraft.world.level.storage.TagValueInput;
 import net.minecraft.world.level.storage.TagValueOutput;
 import net.minecraft.world.level.storage.ValueInput;
 import net.minecraft.world.phys.AABB;
+import net.wechandoit.ArrayConstants;
 import org.slf4j.Logger;
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -400,7 +401,6 @@ public final class ChunkEntitySlices {
 
     private static final class BasicEntityList<E extends Entity> {
 
-        private static final Entity[] EMPTY = new Entity[0];
         private static final int DEFAULT_CAPACITY = 4;
 
         private E[] storage;
@@ -411,7 +411,7 @@ public final class ChunkEntitySlices {
         }
 
         public BasicEntityList(final int cap) {
-            this.storage = (E[])(cap <= 0 ? EMPTY : new Entity[cap]);
+            this.storage = (E[])(cap <= 0 ? ArrayConstants.emptyEntityArray : new Entity[cap]); // SW
         }
 
         public boolean isEmpty() {
@@ -423,7 +423,7 @@ public final class ChunkEntitySlices {
         }
 
         private void resize() {
-            if (this.storage == EMPTY) {
+            if (this.storage == ArrayConstants.emptyEntityArray) { // SW - reduce array allocations
                 this.storage = (E[])new Entity[DEFAULT_CAPACITY];
             } else {
                 this.storage = Arrays.copyOf(this.storage, this.storage.length * 2);
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/server/ServerEntityLookup.java b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/server/ServerEntityLookup.java
index f312a7f5b1b2a777ab36b94ce7cbf3873f5e88bc..d3ac98bbf172c59339dd122cfcce1440da893ca9 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/server/ServerEntityLookup.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/server/ServerEntityLookup.java
@@ -16,13 +16,12 @@ import net.minecraft.server.level.TicketType;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.projectile.throwableitemprojectile.ThrownEnderpearl;
 import net.minecraft.world.level.entity.LevelCallback;
+import net.wechandoit.ArrayConstants;
 
 public final class ServerEntityLookup extends EntityLookup {
 
-    private static final Entity[] EMPTY_ENTITY_ARRAY = new Entity[0];
-
     private final ServerLevel serverWorld;
-    public final ReferenceList<Entity> trackerEntities = new ReferenceList<>(EMPTY_ENTITY_ARRAY); // Moonrise - entity tracker
+    public final ReferenceList<Entity> trackerEntities = new ReferenceList<>(ArrayConstants.emptyEntityArray); // Moonrise - entity tracker + SW - reduce array allocations
 
     // Vanilla does not increment ticket timeouts if the chunk is progressing in generation. They made this change in 1.21.6 so that the ender pearl
     // ticket does not expire if the chunk fails to generate before the timeout expires. Rather than blindly adjusting the entire system behavior
diff --git a/net/minecraft/nbt/ByteArrayTag.java b/net/minecraft/nbt/ByteArrayTag.java
index 9f225fc9c81092d1e593acbf90e1b46141bc3dd0..34ea69af642b46171888ea7728843a0261e767ba 100644
--- a/net/minecraft/nbt/ByteArrayTag.java
+++ b/net/minecraft/nbt/ByteArrayTag.java
@@ -5,6 +5,8 @@ import java.io.DataOutput;
 import java.io.IOException;
 import java.util.Arrays;
 import java.util.Optional;
+
+import net.wechandoit.ArrayConstants;
 import org.apache.commons.lang3.ArrayUtils;
 
 public final class ByteArrayTag implements CollectionTag {
@@ -144,7 +146,7 @@ public final class ByteArrayTag implements CollectionTag {
 
     @Override
     public void clear() {
-        this.data = new byte[0];
+        this.data = ArrayConstants.emptyByteArray; // SW - reduce array allocations
     }
 
     @Override
diff --git a/net/minecraft/nbt/IntArrayTag.java b/net/minecraft/nbt/IntArrayTag.java
index ea1fe06a635601ef1e79518fb4ed3165b3da8360..b9edc628f8447cd9d4fd3abdcc178e258b27f672 100644
--- a/net/minecraft/nbt/IntArrayTag.java
+++ b/net/minecraft/nbt/IntArrayTag.java
@@ -5,6 +5,8 @@ import java.io.DataOutput;
 import java.io.IOException;
 import java.util.Arrays;
 import java.util.Optional;
+
+import net.wechandoit.ArrayConstants;
 import org.apache.commons.lang3.ArrayUtils;
 
 public final class IntArrayTag implements CollectionTag {
@@ -151,7 +153,7 @@ public final class IntArrayTag implements CollectionTag {
 
     @Override
     public void clear() {
-        this.data = new int[0];
+        this.data = ArrayConstants.emptyIntArray; // SW - reduce array allocations
     }
 
     @Override
diff --git a/net/minecraft/nbt/LongArrayTag.java b/net/minecraft/nbt/LongArrayTag.java
index d92f95dcb12f2deb734f71262a30bc0021af6cc5..23706619d19f9c09c160c75cd00016452eec2614 100644
--- a/net/minecraft/nbt/LongArrayTag.java
+++ b/net/minecraft/nbt/LongArrayTag.java
@@ -5,6 +5,8 @@ import java.io.DataOutput;
 import java.io.IOException;
 import java.util.Arrays;
 import java.util.Optional;
+
+import net.wechandoit.ArrayConstants;
 import org.apache.commons.lang3.ArrayUtils;
 
 public final class LongArrayTag implements CollectionTag {
@@ -150,7 +152,7 @@ public final class LongArrayTag implements CollectionTag {
 
     @Override
     public void clear() {
-        this.data = new long[0];
+        this.data = ArrayConstants.emptyLongArray; // SW - reduce array allocations
     }
 
     @Override
diff --git a/net/minecraft/network/CipherBase.java b/net/minecraft/network/CipherBase.java
index 5a1fcd0a552dfcebff351c408a0403a8017930c1..3693cd7a9bf2bb6adb7d6240d909b49d1cd1903b 100644
--- a/net/minecraft/network/CipherBase.java
+++ b/net/minecraft/network/CipherBase.java
@@ -2,13 +2,14 @@ package net.minecraft.network;
 
 import io.netty.buffer.ByteBuf;
 import io.netty.channel.ChannelHandlerContext;
+import net.wechandoit.ArrayConstants;
 import javax.crypto.Cipher;
 import javax.crypto.ShortBufferException;
 
 public class CipherBase {
     private final Cipher cipher;
-    private byte[] heapIn = new byte[0];
-    private byte[] heapOut = new byte[0];
+    private byte[] heapIn = ArrayConstants.emptyByteArray; // SW - reduce array allocations
+    private byte[] heapOut = ArrayConstants.emptyByteArray; // SW - reduce array allocations
 
     protected CipherBase(final Cipher cipher) {
         this.cipher = cipher;
diff --git a/net/minecraft/network/chat/contents/TranslatableContents.java b/net/minecraft/network/chat/contents/TranslatableContents.java
index 35c62bc99962feed075a02e3706496199b1acc64..66f1f08385e369bca81142c0013a5942cfe3cafb 100644
--- a/net/minecraft/network/chat/contents/TranslatableContents.java
+++ b/net/minecraft/network/chat/contents/TranslatableContents.java
@@ -25,10 +25,11 @@ import net.minecraft.network.chat.MutableComponent;
 import net.minecraft.network.chat.ResolutionContext;
 import net.minecraft.network.chat.Style;
 import net.minecraft.util.ExtraCodecs;
+import net.wechandoit.ArrayConstants;
 import org.jspecify.annotations.Nullable;
 
 public class TranslatableContents implements ComponentContents {
-    public static final Object[] NO_ARGS = new Object[0];
+    public static final Object[] NO_ARGS = ArrayConstants.emptyObjectArray; // SW - reduce array allocations
     private static final Codec<Object> PRIMITIVE_ARG_CODEC = ExtraCodecs.JAVA.validate(TranslatableContents::filterAllowedArguments);
     private static final Codec<Object> ARG_CODEC = Codec.either(PRIMITIVE_ARG_CODEC, ComponentSerialization.CODEC)
         .xmap(
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index c0873f9cec8b5877157b7f8803b617465cb0876b..d8084dcdb79b64feea6590f6e989b2b267a8fe38 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -180,6 +180,7 @@ import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 import net.minecraft.world.ticks.LevelTicks;
 import net.minecraft.world.waypoints.WaypointTransmitter;
+import net.wechandoit.ArrayConstants;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
@@ -1400,8 +1401,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
 
     public static List<Entity> getCurrentlyTickingEntities() {
         Entity ticking = currentlyTickingEntity.get();
-        List<Entity> ret = java.util.Arrays.asList(ticking == null ? new Entity[0] : new Entity[] { ticking });
-
+        List<Entity> ret = java.util.Arrays.asList(ticking == null ? ArrayConstants.emptyEntityArray : new Entity[] { ticking }); // SW - reduce array allocations
         return ret;
     }
     // Paper end - log detailed entity tick information
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 06b938013f9dbee4fb3b86979d5a42f28a0ea33e..da44a53cae8e81460a54bfd054f09960ab144c03 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1229,7 +1229,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                     this.getInventory().getNonEquipmentItems().set(i, net.minecraft.world.item.ItemStack.EMPTY);
                 }
             }
-            for (final EquipmentSlot value : EquipmentSlot.VALUES) {
+            for (final EquipmentSlot value : EquipmentSlot.VALUES_ARRAY) {
                 if (this.getInventory().equipment.has(value) && !shouldKeepDeathEventItem(event, this.getInventory().equipment.get(value))) {
                     this.getInventory().equipment.set(value, net.minecraft.world.item.ItemStack.EMPTY);
                 }
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index fe6e148322e920989ac3ffc3c10de95d782265f9..6508870f5b719e910d4ec19ef6586e3e1253a1f4 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -160,16 +160,7 @@ import net.minecraft.util.Util;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.effect.MobEffects;
-import net.minecraft.world.entity.Avatar;
-import net.minecraft.world.entity.Entity;
-import net.minecraft.world.entity.EquipmentSlot;
-import net.minecraft.world.entity.ExperienceOrb;
-import net.minecraft.world.entity.HasCustomInventoryScreen;
-import net.minecraft.world.entity.LivingEntity;
-import net.minecraft.world.entity.MoverType;
-import net.minecraft.world.entity.PlayerRideableJumping;
-import net.minecraft.world.entity.PositionMoveRotation;
-import net.minecraft.world.entity.Relative;
+import net.minecraft.world.entity.*; // SW
 import net.minecraft.world.entity.item.ItemEntity;
 import net.minecraft.world.entity.player.ChatVisiblity;
 import net.minecraft.world.entity.player.Inventory;
@@ -2952,7 +2943,7 @@ public class ServerGamePacketListenerImpl
                             // SPIGOT-7136 - Allays
                             if (target instanceof net.minecraft.world.entity.animal.allay.Allay || target instanceof net.minecraft.world.entity.animal.equine.AbstractHorse) { // Paper - Fix horse armor desync
                                 ServerGamePacketListenerImpl.this.send(new net.minecraft.network.protocol.game.ClientboundSetEquipmentPacket(
-                                    target.getId(), java.util.Arrays.stream(net.minecraft.world.entity.EquipmentSlot.values())
+                                    target.getId(), java.util.Arrays.stream(EquipmentSlot.VALUES_ARRAY) // SW - reduce array allocations
                                     .map((slot) -> com.mojang.datafixers.util.Pair.of(slot, ((LivingEntity) target).getItemBySlot(slot).copy()))
                                     .collect(Collectors.toList()), true)); // Paper - sanitize
                                 player.containerMenu.sendAllDataToRemote();
diff --git a/net/minecraft/server/players/StoredUserList.java b/net/minecraft/server/players/StoredUserList.java
index d97d79e65c672e31e408f7627189ab65f5aedc2c..41a87daeab5e3977e06ddfe465f75c4d07b38b71 100644
--- a/net/minecraft/server/players/StoredUserList.java
+++ b/net/minecraft/server/players/StoredUserList.java
@@ -21,6 +21,7 @@ import java.util.Objects;
 import net.minecraft.server.notifications.NotificationService;
 import net.minecraft.util.GsonHelper;
 import net.minecraft.util.Util;
+import net.wechandoit.ArrayConstants;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
@@ -96,7 +97,7 @@ public abstract class StoredUserList<K, V extends StoredUserEntry<K>> {
     }
 
     public String[] getUserList() {
-        return this.map.keySet().toArray(new String[0]);
+        return this.map.keySet().toArray(ArrayConstants.emptyStringArray);
     }
 
     public boolean isEmpty() {
diff --git a/net/minecraft/util/ZeroBitStorage.java b/net/minecraft/util/ZeroBitStorage.java
index eb2fa32cff6824c14f865c8731df7d082122f62b..11f3811b96ffdeda0745608ad484d15c7e949a66 100644
--- a/net/minecraft/util/ZeroBitStorage.java
+++ b/net/minecraft/util/ZeroBitStorage.java
@@ -2,10 +2,11 @@ package net.minecraft.util;
 
 import java.util.Arrays;
 import java.util.function.IntConsumer;
+import net.wechandoit.ArrayConstants;
 import org.apache.commons.lang3.Validate;
 
 public class ZeroBitStorage implements BitStorage {
-    public static final long[] RAW = new long[0];
+    public static final long[] RAW = ArrayConstants.emptyLongArray; // SW - reduce array allocations
     private final int size;
 
     public ZeroBitStorage(final int size) {
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 9cd46a8619e575a865f4d00ce71d249c9b028b2f..9344a50660887e37b51a5d9215d121b49db29572 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -3179,7 +3179,7 @@ public abstract class Entity
     }
 
     private boolean attemptToShearEquipment(final Player player, final InteractionHand hand, final ItemStack heldItem, final Mob target) {
-        for (EquipmentSlot slot : EquipmentSlot.VALUES) {
+        for (EquipmentSlot slot : EquipmentSlot.VALUES_ARRAY) {
             ItemStack itemStack = target.getItemBySlot(slot);
             Equippable equippable = itemStack.get(DataComponents.EQUIPPABLE);
             if (equippable != null
diff --git a/net/minecraft/world/entity/EquipmentSlot.java b/net/minecraft/world/entity/EquipmentSlot.java
index 33d296e826093ce3afc9f1edaa6ccbd6c25a5a7d..bf9787e82b408961ea6eae9529aa35218482b63e 100644
--- a/net/minecraft/world/entity/EquipmentSlot.java
+++ b/net/minecraft/world/entity/EquipmentSlot.java
@@ -20,9 +20,12 @@ public enum EquipmentSlot implements StringRepresentable {
     SADDLE(EquipmentSlot.Type.SADDLE, 0, 1, 7, "saddle");
 
     public static final int NO_COUNT_LIMIT = 0;
-    public static final List<EquipmentSlot> VALUES = List.of(values());
-    public static final IntFunction<EquipmentSlot> BY_ID = ByIdMap.continuous(s -> s.id, values(), ByIdMap.OutOfBoundsStrategy.ZERO);
-    public static final StringRepresentable.EnumCodec<EquipmentSlot> CODEC = StringRepresentable.fromEnum(EquipmentSlot::values);
+    // SW start - reduce array allocations
+    public static final EquipmentSlot[] VALUES_ARRAY = values();
+    public static final List<EquipmentSlot> VALUES = List.of(VALUES_ARRAY);
+    public static final IntFunction<EquipmentSlot> BY_ID = ByIdMap.continuous(equipmentSlot -> equipmentSlot.id, VALUES_ARRAY, ByIdMap.OutOfBoundsStrategy.ZERO);
+    public static final StringRepresentable.EnumCodec<EquipmentSlot> CODEC = StringRepresentable.fromEnum(() -> VALUES_ARRAY);
+    // SW end - reduce array allocations
     public static final StreamCodec<ByteBuf, EquipmentSlot> STREAM_CODEC = ByteBufCodecs.idMapper(BY_ID, s -> s.id);
     private final EquipmentSlot.Type type;
     private final int index;
diff --git a/net/minecraft/world/entity/EquipmentSlotGroup.java b/net/minecraft/world/entity/EquipmentSlotGroup.java
index ea365265e40c0efa386f47e3fa45bc829ef552fa..49e1f62c44398c2ba3536633d41af497e9591340 100644
--- a/net/minecraft/world/entity/EquipmentSlotGroup.java
+++ b/net/minecraft/world/entity/EquipmentSlotGroup.java
@@ -24,6 +24,7 @@ public enum EquipmentSlotGroup implements StringRepresentable, Iterable<Equipmen
     BODY(9, "body", EquipmentSlot.BODY),
     SADDLE(10, "saddle", EquipmentSlot.SADDLE);
 
+    public static final EquipmentSlotGroup[] VALUES_ARRAY = EquipmentSlotGroup.values(); // SW - reduce array allocations
     public static final IntFunction<EquipmentSlotGroup> BY_ID = ByIdMap.continuous(s -> s.id, values(), ByIdMap.OutOfBoundsStrategy.ZERO);
     public static final Codec<EquipmentSlotGroup> CODEC = StringRepresentable.fromEnum(EquipmentSlotGroup::values);
     public static final StreamCodec<ByteBuf, EquipmentSlotGroup> STREAM_CODEC = ByteBufCodecs.idMapper(BY_ID, s -> s.id);
diff --git a/net/minecraft/world/entity/EquipmentTable.java b/net/minecraft/world/entity/EquipmentTable.java
index 4b018c64d3123bb7a0687491d2ac3a535302890c..f71cb65d2e8c6a6030b2fc0501bb7537cd9ff241 100644
--- a/net/minecraft/world/entity/EquipmentTable.java
+++ b/net/minecraft/world/entity/EquipmentTable.java
@@ -35,7 +35,7 @@ public record EquipmentTable(ResourceKey<LootTable> lootTable, Map<EquipmentSlot
     }
 
     private static Map<EquipmentSlot, Float> createForAllSlots(final float dropChance) {
-        return createForAllSlots(List.of(EquipmentSlot.values()), dropChance);
+        return createForAllSlots(List.of(EquipmentSlot.VALUES_ARRAY), dropChance); // SW - reduce array allocations
     }
 
     private static Map<EquipmentSlot, Float> createForAllSlots(final List<EquipmentSlot> slots, final float dropChance) {
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 6a0ebbdfcc0bc07915574c1cf87463e82213ca7e..05ecb567b0c11182eeef0ed31dccbab1f418d363 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -3579,7 +3579,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         Map<org.bukkit.inventory.EquipmentSlot, io.papermc.paper.event.entity.EntityEquipmentChangedEvent.EquipmentChange> equipmentChanges = null;
         // Paper end - EntityEquipmentChangedEvent
 
-        for (EquipmentSlot slot : EquipmentSlot.VALUES) {
+        for (EquipmentSlot slot : EquipmentSlot.VALUES_ARRAY) {
             ItemStack previous = this.lastEquipmentItems.get(slot);
             ItemStack current = this.getItemBySlot(slot);
             if (this.equipmentHasChanged(previous, current)) {
@@ -3862,7 +3862,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
     protected boolean canGlide() {
         if (!this.onGround() && !this.isPassenger() && !this.hasEffect(MobEffects.LEVITATION)) {
-            for (EquipmentSlot slot : EquipmentSlot.VALUES) {
+            for (EquipmentSlot slot : EquipmentSlot.VALUES_ARRAY) {
                 if (canGlideUsing(this.getItemBySlot(slot), slot)) {
                     return true;
                 }
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 5bbb5cbb51953399c7bce547cd6968eb3c6bd130..bed371bf6dcc4320d81fcc157623941172332e08 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -364,7 +364,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
         if (this.xpReward > 0) {
             int result = this.xpReward;
 
-            for (EquipmentSlot slot : EquipmentSlot.VALUES) {
+            for (EquipmentSlot slot : EquipmentSlot.VALUES_ARRAY) {
                 if (slot.canIncreaseExperience()) {
                     ItemStack item = this.getItemBySlot(slot);
                     if (!item.isEmpty() && this.dropChances.byEquipment(slot) <= 1.0F) {
@@ -1044,7 +1044,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
     protected void dropCustomDeathLoot(final ServerLevel level, final DamageSource source, final boolean killedByPlayer) {
         super.dropCustomDeathLoot(level, source, killedByPlayer);
 
-        for (EquipmentSlot slot : EquipmentSlot.VALUES) {
+        for (EquipmentSlot slot : EquipmentSlot.VALUES_ARRAY) {
             if (this.shouldSkipLoot(slot)) continue; // Paper
             ItemStack itemStack = this.getItemBySlot(slot);
             float dropChance = this.dropChances.byEquipment(slot);
@@ -1088,7 +1088,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
     public Set<EquipmentSlot> dropPreservedEquipment(final ServerLevel level, final Predicate<ItemStack> shouldDrop) {
         Set<EquipmentSlot> slotsPreventedFromDropping = new HashSet<>();
 
-        for (EquipmentSlot slot : EquipmentSlot.VALUES) {
+        for (EquipmentSlot slot : EquipmentSlot.VALUES_ARRAY) {
             ItemStack itemStack = this.getItemBySlot(slot);
             if (!itemStack.isEmpty()) {
                 if (!shouldDrop.test(itemStack)) {
@@ -1218,7 +1218,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
     protected void populateDefaultEquipmentEnchantments(final ServerLevelAccessor level, final RandomSource random, final DifficultyInstance localDifficulty) {
         this.enchantSpawnedWeapon(level, random, localDifficulty);
 
-        for (EquipmentSlot slot : EquipmentSlot.VALUES) {
+        for (EquipmentSlot slot : EquipmentSlot.VALUES_ARRAY) {
             if (slot.getType() == EquipmentSlot.Type.HUMANOID_ARMOR) {
                 this.enchantSpawnedArmor(level, random, slot, localDifficulty);
             }
@@ -1639,7 +1639,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
     protected void removeAfterChangingDimensions() {
         super.removeAfterChangingDimensions();
 
-        for (EquipmentSlot slot : EquipmentSlot.VALUES) {
+        for (EquipmentSlot slot : EquipmentSlot.VALUES_ARRAY) {
             ItemStack itemStack = this.getItemBySlot(slot);
             if (!itemStack.isEmpty()) {
                 itemStack.setCount(0);
diff --git a/net/minecraft/world/entity/decoration/ArmorStand.java b/net/minecraft/world/entity/decoration/ArmorStand.java
index fc4fce0b1d57a62c1e5c80a5b91c2b1a77d4bace..27fc32f8edc60554ab3732280b4483dd19616555 100644
--- a/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -458,7 +458,7 @@ public class ArmorStand extends LivingEntity {
         this.playBrokenSound();
         // this.dropAllDeathLoot(level, source); // CraftBukkit - moved down
 
-        for (EquipmentSlot slot : EquipmentSlot.VALUES) {
+        for (EquipmentSlot slot : EquipmentSlot.VALUES_ARRAY) {
             ItemStack itemStack = this.equipment.get(slot); // Paper - move equipment removal past event call
             if (!itemStack.isEmpty() && !EnchantmentHelper.has(itemStack, EnchantmentEffectComponents.PREVENT_EQUIPMENT_DROP)) {
                 this.drops.add(new DefaultDrop(itemStack, stack -> Block.popResource(this.level(), this.blockPosition().above(), stack))); // CraftBukkit - add to drops // Paper - Restore vanilla drops behavior; mirror so we can destroy it later - though this call site was safe & spawn drops correctly}
@@ -467,7 +467,7 @@ public class ArmorStand extends LivingEntity {
         // Paper start - move equipment removal past event call
         org.bukkit.event.entity.EntityDeathEvent event = this.dropAllDeathLoot(level, source);
         if (!event.isCancelled()) {
-            for (EquipmentSlot slot : EquipmentSlot.VALUES) {
+            for (EquipmentSlot slot : EquipmentSlot.VALUES_ARRAY) {
                 this.equipment.set(slot, ItemStack.EMPTY);
             }
         }
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index b8c40b7e9a52211181bcd3d8e8c6041b6319d727..800d2d87d80827375c6bcfc405b6115c9938fd82 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -341,7 +341,7 @@ public abstract class Player extends Avatar implements ContainerUser {
     }
 
     private boolean isEquipped(final Item item) {
-        for (EquipmentSlot slot : EquipmentSlot.VALUES) {
+        for (EquipmentSlot slot : EquipmentSlot.VALUES_ARRAY) {
             ItemStack itemStack = this.getItemBySlot(slot);
             Equippable equippable = itemStack.get(DataComponents.EQUIPPABLE);
             if (itemStack.is(item) && equippable != null && equippable.slot() == slot) {
diff --git a/net/minecraft/world/item/ItemStack.java b/net/minecraft/world/item/ItemStack.java
index 54d67c847c1de9ee216cedb0527136f50223ea2d..1060bf4fba9aac6b151ccd0daef49a5d3ddff31e 100644
--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -1195,7 +1195,7 @@ public final class ItemStack implements DataComponentHolder, ItemInstance {
 
     private void addAttributeTooltips(final Consumer<Component> consumer, final TooltipDisplay display, final @Nullable Player player) {
         if (display.shows(DataComponents.ATTRIBUTE_MODIFIERS)) {
-            for (EquipmentSlotGroup slot : EquipmentSlotGroup.values()) {
+            for (EquipmentSlotGroup slot : EquipmentSlotGroup.VALUES_ARRAY) {
                 MutableBoolean first = new MutableBoolean(true);
                 this.forEachModifier(slot, (attribute, modifier, tooltip) -> {
                     if (tooltip != ItemAttributeModifiers.Display.hidden()) {
diff --git a/net/minecraft/world/item/crafting/ShapedRecipePattern.java b/net/minecraft/world/item/crafting/ShapedRecipePattern.java
index a8501a16350e98f6bf28482bde0b68d55ca6d222..f3a71b60b633d754ed5ce4d97630e73f3e668641 100644
--- a/net/minecraft/world/item/crafting/ShapedRecipePattern.java
+++ b/net/minecraft/world/item/crafting/ShapedRecipePattern.java
@@ -18,6 +18,7 @@ import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.util.ExtraCodecs;
 import net.minecraft.util.Util;
 import net.minecraft.world.item.ItemStack;
+import net.wechandoit.ArrayConstants;
 
 public final class ShapedRecipePattern {
     private static final int MAX_SIZE = 3;
@@ -121,7 +122,7 @@ public final class ShapedRecipePattern {
         }
 
         if (pattern.size() == bottom) {
-            return new String[0];
+            return ArrayConstants.emptyStringArray; // SW - reduce array allocations
         } else {
             String[] result = new String[pattern.size() - bottom - top];
 
diff --git a/net/minecraft/world/item/enchantment/Enchantment.java b/net/minecraft/world/item/enchantment/Enchantment.java
index 0d981b86fa7ee8713e2a419fc2f377defbc2ed9c..a71d2011f25700a230cab99074de005fce29f8f7 100644
--- a/net/minecraft/world/item/enchantment/Enchantment.java
+++ b/net/minecraft/world/item/enchantment/Enchantment.java
@@ -107,7 +107,7 @@ public record Enchantment(Component description, Enchantment.EnchantmentDefiniti
     public Map<EquipmentSlot, ItemStack> getSlotItems(final LivingEntity entity) {
         Map<EquipmentSlot, ItemStack> itemStacks = Maps.newEnumMap(EquipmentSlot.class);
 
-        for (EquipmentSlot slot : EquipmentSlot.VALUES) {
+        for (EquipmentSlot slot : EquipmentSlot.VALUES_ARRAY) {
             if (this.matchingSlot(slot)) {
                 ItemStack itemStack = entity.getItemBySlot(slot);
                 if (!itemStack.isEmpty()) {
diff --git a/net/minecraft/world/item/enchantment/EnchantmentHelper.java b/net/minecraft/world/item/enchantment/EnchantmentHelper.java
index d5f749fd0745555fa91e184abea0c40ab873c182..ba9ef474d31579033df6709d01a89dbaefcf1be6 100644
--- a/net/minecraft/world/item/enchantment/EnchantmentHelper.java
+++ b/net/minecraft/world/item/enchantment/EnchantmentHelper.java
@@ -157,7 +157,7 @@ public class EnchantmentHelper {
     }
 
     private static void runIterationOnEquipment(final LivingEntity owner, final EnchantmentHelper.EnchantmentInSlotVisitor method) {
-        for (EquipmentSlot slot : EquipmentSlot.VALUES) {
+        for (EquipmentSlot slot : EquipmentSlot.VALUES_ARRAY) {
             runIterationOnItem(owner.getItemBySlot(slot), slot, owner, method);
         }
     }
@@ -495,7 +495,7 @@ public class EnchantmentHelper {
     ) {
         List<EnchantedItemInUse> items = new ArrayList<>();
 
-        for (EquipmentSlot slot : EquipmentSlot.VALUES) {
+        for (EquipmentSlot slot : EquipmentSlot.VALUES_ARRAY) {
             ItemStack item = source.getItemBySlot(slot);
             if (predicate.test(item)) {
                 ItemEnchantments enchantments = item.getOrDefault(DataComponents.ENCHANTMENTS, ItemEnchantments.EMPTY);
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index e1fd30fc9b17d241c421c66114c9611b2979053e..b957978d9ee0588b9790d736f6ee5c91e4e9409c 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -89,6 +89,7 @@ import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.VoxelShape;
 import net.minecraft.world.scores.Scoreboard;
+import net.wechandoit.ArrayConstants;
 import org.apache.commons.lang3.mutable.MutableBoolean;
 import org.bukkit.craftbukkit.util.CraftNamespacedKey;
 import org.jspecify.annotations.Nullable;
@@ -1861,7 +1862,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     public org.bukkit.entity.Entity[] getChunkEntities(int chunkX, int chunkZ) {
         ca.spottedleaf.moonrise.patches.chunk_system.level.entity.ChunkEntitySlices slices = ((ServerLevel)this).moonrise$getEntityLookup().getChunk(chunkX, chunkZ);
         if (slices == null) {
-            return new org.bukkit.entity.Entity[0];
+            return ArrayConstants.emptyBukkitEntityArray; // SW - return empty array
         }
 
         List<org.bukkit.entity.Entity> ret = new java.util.ArrayList<>();
@@ -1872,7 +1873,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
             }
         }
 
-        return ret.toArray(new org.bukkit.entity.Entity[0]);
+        return ret.toArray(ArrayConstants.emptyBukkitEntityArray); // SW - return empty array
     }
     // Paper end - rewrite chunk system
 
diff --git a/net/minecraft/world/level/block/ComposterBlock.java b/net/minecraft/world/level/block/ComposterBlock.java
index 3eb65e5ed0241330ebde3e05f7865e294d4b48e7..57c10235f15f22adb4b5e03b5d9bf368bfb3cb7b 100644
--- a/net/minecraft/world/level/block/ComposterBlock.java
+++ b/net/minecraft/world/level/block/ComposterBlock.java
@@ -39,6 +39,7 @@ import net.minecraft.world.phys.shapes.BooleanOp;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import net.wechandoit.ArrayConstants;
 import org.jspecify.annotations.Nullable;
 
 public class ComposterBlock extends Block implements WorldlyContainerHolder {
@@ -434,7 +435,7 @@ public class ComposterBlock extends Block implements WorldlyContainerHolder {
 
         @Override
         public int[] getSlotsForFace(final Direction direction) {
-            return new int[0];
+            return ArrayConstants.emptyIntArray; // SW - reduce array allocations
         }
 
         @Override
@@ -469,7 +470,7 @@ public class ComposterBlock extends Block implements WorldlyContainerHolder {
 
         @Override
         public int[] getSlotsForFace(final Direction direction) {
-            return direction == Direction.UP ? new int[]{0} : new int[0];
+            return direction == Direction.UP ? ArrayConstants.zeroSingletonIntArray : ArrayConstants.emptyIntArray;
         }
 
         @Override
@@ -520,7 +521,7 @@ public class ComposterBlock extends Block implements WorldlyContainerHolder {
 
         @Override
         public int[] getSlotsForFace(final Direction direction) {
-            return direction == Direction.DOWN ? new int[]{0} : new int[0];
+            return direction == Direction.DOWN ? ArrayConstants.zeroSingletonIntArray : ArrayConstants.emptyIntArray;
         }
 
         @Override
diff --git a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
index f41b01bfa9443b46f9ca0d33d1b03280f0214b93..68c0a9cecabb0032c32784f7e0efe24becb76648 100644
--- a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -38,6 +38,7 @@ import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.storage.ValueInput;
 import net.minecraft.world.level.storage.ValueOutput;
 import net.minecraft.world.phys.Vec3;
+import net.wechandoit.ArrayConstants;
 import org.jspecify.annotations.Nullable;
 
 public abstract class AbstractFurnaceBlockEntity extends BaseContainerBlockEntity implements WorldlyContainer, StackedContentsCompatible, RecipeCraftingHolder {
@@ -45,7 +46,7 @@ public abstract class AbstractFurnaceBlockEntity extends BaseContainerBlockEntit
     protected static final int SLOT_FUEL = 1;
     protected static final int SLOT_RESULT = 2;
     public static final int DATA_LIT_TIME = 0;
-    private static final int[] SLOTS_FOR_UP = new int[]{0};
+    private static final int[] SLOTS_FOR_UP = ArrayConstants.zeroSingletonIntArray;
     private static final int[] SLOTS_FOR_DOWN = new int[]{2, 1};
     private static final int[] SLOTS_FOR_SIDES = new int[]{1};
     public static final int DATA_LIT_DURATION = 1;
diff --git a/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java b/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
index 83a3ecf728be1129b07b5ab450861f2154cee92b..2cd8e7c31cb8c88e2e8192dbd40f8c576d6f02e5 100644
--- a/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
+++ b/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
@@ -269,7 +269,7 @@ public class MapItemSavedData extends SavedData {
     }
 
     private static boolean hasMapInvisibilityItemEquipped(final Player player) {
-        for (EquipmentSlot equipmentSlot : EquipmentSlot.values()) {
+        for (EquipmentSlot equipmentSlot : EquipmentSlot.VALUES_ARRAY) { // SW - reduce array allocations
             if (equipmentSlot != EquipmentSlot.MAINHAND
                 && equipmentSlot != EquipmentSlot.OFFHAND
                 && player.getItemBySlot(equipmentSlot).is(ItemTags.MAP_INVISIBILITY_EQUIPMENT)) {
diff --git a/net/wechandoit/ArrayConstants.java b/net/wechandoit/ArrayConstants.java
new file mode 100644
index 0000000000000000000000000000000000000000..063e3130140faf4978d805bbe370c9e5b65d1ebb
--- /dev/null
+++ b/net/wechandoit/ArrayConstants.java
@@ -0,0 +1,20 @@
+// From JettPack/Gale - reduce array allocations
+
+package net.wechandoit;
+
+public final class ArrayConstants {
+
+    private ArrayConstants() {}
+
+    public static final Object[] emptyObjectArray = new Object[0];
+    public static final short[] emptyShortArray = new short[0];
+    public static final int[] emptyIntArray = new int[0];
+    public static final int[] zeroSingletonIntArray = new int[]{0};
+    public static final byte[] emptyByteArray = new byte[0];
+    public static final String[] emptyStringArray = new String[0];
+    public static final long[] emptyLongArray = new long[0];
+    public static final org.bukkit.entity.Entity[] emptyBukkitEntityArray = new org.bukkit.entity.Entity[0];
+    public static final net.minecraft.world.entity.Entity[] emptyEntityArray = new net.minecraft.world.entity.Entity[0];
+    //public static final net.minecraft.server.level.ServerLevel[] emptyServerLevelArray = new net.minecraft.server.level.ServerLevel[0];
+    public static final net.minecraft.tags.TagKey[] emptyTagKeyArray = new net.minecraft.tags.TagKey[0]; // Leaf - Optimize isEyeInFluid
+}
\ No newline at end of file
