From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wechandoit <ethanchan7981@gmail.com>
Date: Thu, 7 May 2026 01:01:00 -0400
Subject: [PATCH] Pluto - Avoid excessive calls to Enum#values


diff --git a/net/minecraft/core/Direction.java b/net/minecraft/core/Direction.java
index af661c06064a27e70fe4522dac5fe7342b9b117e..056ff0f5e03e414700d93799d6a4b0d5ae1c92bb 100644
--- a/net/minecraft/core/Direction.java
+++ b/net/minecraft/core/Direction.java
@@ -56,7 +56,7 @@ public enum Direction implements StringRepresentable, ca.spottedleaf.moonrise.pa
     private final Vec3i normal;
     private final Vec3 normalVec3;
     private final Vector3fc normalVec3f;
-    private static final Direction[] VALUES = values();
+    public static final Direction[] VALUES = values(); // SW: Pluto - Avoid excessive calls to Enum#values
     private static final Direction[] BY_3D_DATA = Arrays.stream(VALUES).sorted(Comparator.comparingInt(d -> d.data3d)).toArray(Direction[]::new);
     private static final Direction[] BY_2D_DATA = Arrays.stream(VALUES)
         .filter(d -> d.getAxis().isHorizontal())
diff --git a/net/minecraft/world/level/redstone/DefaultRedstoneWireEvaluator.java b/net/minecraft/world/level/redstone/DefaultRedstoneWireEvaluator.java
index bfa6f2420677e4e55c22b473ea23e91e159ed92f..b9b33c76a6b93c97a6310cabc301e7370315714d 100644
--- a/net/minecraft/world/level/redstone/DefaultRedstoneWireEvaluator.java
+++ b/net/minecraft/world/level/redstone/DefaultRedstoneWireEvaluator.java
@@ -34,7 +34,7 @@ public class DefaultRedstoneWireEvaluator extends RedstoneWireEvaluator {
             Set<BlockPos> toUpdate = Sets.newHashSet();
             toUpdate.add(pos);
 
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // SW: Pluto - Avoid excessive calls to Enum#values
                 toUpdate.add(pos.relative(direction));
             }
 
diff --git a/net/minecraft/world/level/redstone/Orientation.java b/net/minecraft/world/level/redstone/Orientation.java
index f71dc76e61103a80d5d8d574bfb0648fef3f5abc..91050db9a5d4b800f71964f8aa4757af738d4bcf 100644
--- a/net/minecraft/world/level/redstone/Orientation.java
+++ b/net/minecraft/world/level/redstone/Orientation.java
@@ -136,7 +136,7 @@ public class Orientation {
             self.withSideBias.put(sideBias, generateContext(new Orientation(self.up, self.front, sideBias), lookup));
         }
 
-        for (Direction facing : Direction.values()) {
+            for (Direction facing : Direction.VALUES) { // SW: Pluto - Avoid excessive calls to Enum#values
             Direction up = self.up;
             if (facing == self.up) {
                 up = self.front.getOpposite();
@@ -149,7 +149,7 @@ public class Orientation {
             self.withFront.put(facing, generateContext(new Orientation(up, facing, self.sideBias), lookup));
         }
 
-        for (Direction facing : Direction.values()) {
+            for (Direction facing : Direction.VALUES) { // SW: Pluto - Avoid excessive calls to Enum#values
             Direction front = self.front;
             if (facing == self.front) {
                 front = self.up.getOpposite();
